tests/integration/consumer/test_consumer_upgrade_downgrade.py

```diff
@@ -0,0 +1,131 @@
+
+
+# def produce_messages(producer, topic, partitions, num_messages):
+#     for i in range(num_messages):
+#         key = "key-{}".format(i)
+#         value = "value-{}".format(i)
+#         partition = i % partitions
+#         producer.produce(topic, key=key, value=value, partition=partition)
+#     producer.flush()
+
+
+def check_consumer(kafka_cluster, consumers, admin_client, topic, expected_protocol):
+    total_msg_read = 0
+    while len(consumers[-1].assignment()) != number_of_partitions // len(consumers):
+        for consumer in consumers:
+            consumer.poll(0.1)
+
+    for consumer in consumers:
+        assert len(consumer.assignment()) == number_of_partitions // len(consumers)
+
+    assert get_group_protocol_type(admin_client, topic) == expected_protocol
+
+    # Produce some messages to the topic
+    kafka_cluster.seed_topic(topic)
+    list_offsets(admin_client, topic, number_of_partitions)
+
+    while total_msg_read < 100:
+        for consumer in consumers:
+            # Poll for messages
+            msg = consumer.poll(0.1)
+            if msg is not None:
+                total_msg_read += 1
+    
+    assert total_msg_read == 100, "Expected to read 100 messages, but read {}".format(total_msg_read)
+
+
+def perform_consumer_upgrade_downgrade_test_with_partition_assignment_strategy(kafka_cluster, partition_assignment_strategy):
+    """
+    Test consumer upgrade and downgrade.
+    """
+    topic = kafka_cluster.create_topic_and_wait_propogation(topic_prefix,
+                                                                    {
+                                                                        "num_partitions": number_of_partitions
+                                                                    })
+    admin_client = kafka_cluster.admin()
+
+    # Create a consumer with the latest version
+    consumer_conf = {'group.id': topic,
+                     'auto.offset.reset': 'earliest',
+                     'group.protocol': 'classic'}
+    consumer_conf['partition.assignment.strategy'] = partition_assignment_strategy
+    consumer = kafka_cluster.consumer(consumer_conf)
+    assert consumer is not None
+    consumer.subscribe([topic])
+    check_consumer(kafka_cluster, [consumer], admin_client, topic, ConsumerGroupType.CLASSIC)
+    del consumer_conf['partition.assignment.strategy']
+
+    # Now simulate an upgrade by creating a new consumer with 'consumer' protocol
+    consumer_conf['group.protocol'] = 'consumer'
+    consumer2 = kafka_cluster.consumer(consumer_conf)
+    assert consumer2 is not None
+    consumer2.subscribe([topic])
+    check_consumer(kafka_cluster, [consumer, consumer2], admin_client, topic, ConsumerGroupType.CONSUMER)
+    
+    # Now simulate a downgrade by deleting the second consumer and keeping only 'classic' consumer
+    consumer2.close()
+    check_consumer(kafka_cluster, [consumer], admin_client, topic, ConsumerGroupType.CLASSIC)
+
+    consumer.close()
```

Comment:
This part is limited to two consumers and it's not checking the case where all classic consumers are removed.
What you can do is having more consumer, let's say `consumers_cnt = 5` then, running `check_consumer` after each step:

1. first you add all classic ones, you verify the protocol that is `CLASSIC`
2. one by one you `pop` the first consumer, close it and add a new consumer with `CONSUMER` protocol. you verify the protocol that is `CONSUMER` because it changes on first consumer added.
3. when all consumers are with new protocol you do the same, this time it replaces `consumer` protocol consumers with `classic` consumers. You verify that the protocol is `CONSUMER` until after last replacement where it returns to `CLASSIC`.
4. you close all consumers for next test run.